design-add-and-search-words-data-structure/flynn.cpp

```diff
@@ -0,0 +1,89 @@
+/**
+ * 풀이
+ * - Trie 구조를 활용하여 풀이할 수 있습니다
+ * - wildcard인 '.'에 대한 처리가 필요합니다
+ * 
+ * Big O
+ * - N: 주어지는 문자열 word의 길이
+ * - M: 현재 WordDictionary에 저장되어 있는 TrieNode의 수
+ * 
+ * - void addWord(string word)
+ *   - Time complexity: O(N)
+ *   - Space complexity: O(N)
+ *     - 최악의 경우 word의 모든 문자에 대해 새로운 TrieNode를 추가해야 합니다
+ * 
+ * - bool search(string word)
+ * - bool _search(string word, int idx, TrieNode* node)
+ *   - Time complexity: best O(N), worst O(M) < O(26^N)
+ *     - wildcard 사용 및 기존에 저장된 word의 상태에 따라 달라집니다
+ *   - Space complexity: O(N)
+ *     - _search가 재귀적으로 호출되므로 재귀 호출 스택의 깊이만큼 추가적인 공간이 사용됩니다
+ *     - 재귀 호출 스택의 깊이는 현재 찾는 word의 길이에 선형적으로 비례합니다
+ */
+
+class TrieNode {
+public:
+    array<TrieNode*, 27> links;
+    bool word;
+
+    TrieNode(): word(false) {
+        links.fill(nullptr);
+    }
+};
+
+class WordDictionary {
+public:
+    WordDictionary(): root(new TrieNode()) {}
+    
+    void addWord(string word) {
+        TrieNode* current = root;
+        
+        for (char c : word) {
+            if (current->links[c - 'a'] == nullptr) {
+                current->links[c - 'a'] = new TrieNode();
+            }
+            current = current->links[c - 'a'];
+        }
+
+        current->word = true;
+    }
+    
+    bool search(string word) {
+        return _search(word, 0, root);
+    }
+    
+private:
+    TrieNode* root;
+
+    bool _search(string word, int idx, TrieNode* node) {
```

Comment:
(사소) node 가 null 이면 false 로 반환하는 조건을 `L63` 이전에 두면, `L64`와 `L72`의 if 조건을 뺄 수 있을 것 같습니다!

Reply:
감사합니다 :)
L64, L72를 공통적으로 처리할 수 있는 방법이 잘 떠오르지 않아서 그러는데 예시 코드를 한 번 부탁드려도 될까요?

Reply:
아 네 저는 이렇게 생각했습니다! 성능상 이점이 있다기보다 가독성을 조금 더 생각했다고 봐주시면 될 것 같습니당

```suggestion
    bool _search(string word, int idx, TrieNode* node) {
        if (node == nullptr) return false;
        
        if (word.size() == idx) return node->word; // Added

        char c = word[idx];

        if (c != '.') {
            return _search(word, idx + 1, node->links[c - 'a']);
        } else {
            for (TrieNode* link : node->links) {
                if (_search(word, idx + 1, link)) return true;
            }
            return false;
        }
    }
```

Reply:
오우 감사합니다! ㅎㅎㅎ
이 방식으로 수정하는게 더 좋아보입니다